packages/protocol/contracts/L1/hooks/AssignmentHook.sol

```diff
@@ -0,0 +1,172 @@
+// SPDX-License-Identifier: MIT
+//  _____     _ _         _         _
+// |_   _|_ _(_) |_____  | |   __ _| |__ ___
+//   | |/ _` | | / / _ \ | |__/ _` | '_ (_-<
+//   |_|\__,_|_|_\_\___/ |____\__,_|_.__/__/
+
+pragma solidity ^0.8.20;
+
+import { ERC20Upgradeable } from
+    "lib/openzeppelin-contracts-upgradeable/contracts/token/ERC20/ERC20Upgradeable.sol";
+
+import { AddressResolver } from "../../common/AddressResolver.sol";
+import { EssentialContract } from "../../common/EssentialContract.sol";
+import { Proxied } from "../../common/Proxied.sol";
+import { LibAddress } from "../../libs/LibAddress.sol";
+
+import { TaikoData } from "../TaikoData.sol";
+import { TaikoToken } from "../TaikoToken.sol";
+
+import { IHook } from "./IHook.sol";
+
+/// @title AssignmentHook
+/// A hook that handles prover assignment varification and fee processing.
+contract AssignmentHook is EssentialContract, IHook {
+    using LibAddress for address;
+
+    // Max gas paying the prover. This should be large enough to prevent the
+    // worst cases, usually block proposer shall be aware the risks and only
+    // choose provers that cannot consume too much gas when receiving Ether.
+    uint256 public constant MAX_GAS_PAYING_PROVER = 200_000;
+
+    struct ProverAssignment {
+        address feeToken;
+        uint64 expiry;
+        uint64 maxBlockId;
+        uint64 maxProposedIn;
+        bytes32 metaHash;
+        TaikoData.TierFee[] tierFees;
+        bytes signature;
+    }
+
+    struct Input {
+        ProverAssignment assignment;
+        uint256 tip;
+    }
+
+    error HOOK_ASSIGNMENT_EXPIRED();
+    error HOOK_ASSIGNMENT_INVALID_SIG();
+    error HOOK_ASSIGNMENT_INSUFFICIENT_FEE();
+    error HOOK_TIER_NOT_FOUND();
+
+    function init(address _addressManager) external initializer {
+        EssentialContract._init(_addressManager);
+    }
+
+    function onBlockProposed(
+        TaikoData.Block memory blk,
+        TaikoData.BlockMetadata memory meta,
+        bytes memory data
+    )
+        external
+        payable
+        nonReentrant
+        onlyFromNamed("taiko")
+    {
+        Input memory input = abi.decode(data, (Input));
+        ProverAssignment memory assignment = input.assignment;
+
+        // Check assignment validity
+        if (
+            block.timestamp > assignment.expiry
+                || assignment.metaHash != 0 && blk.metaHash != assignment.metaHash
+                || assignment.maxBlockId != 0 && meta.id > assignment.maxBlockId
+                || assignment.maxProposedIn != 0
+                    && block.number > assignment.maxProposedIn
+        ) {
+            revert HOOK_ASSIGNMENT_EXPIRED();
+        }
+
+        // Hash the assignment with the blobHash, this hash will be signed by
+        // the prover, therefore, we add a string as a prefix.
+        bytes32 hash = hashAssignment(assignment, address(this), meta.blobHash);
```

Comment:
According to the function declaration (`hashAssignment()`) i think it shall be `msg.sender`
```suggestion
        bytes32 hash = hashAssignment(assignment, msg.sender, meta.blobHash);
```

Reply:
I think both will work.